client/src/document/toolbar/flaws.tsx

```diff
@@ -107,61 +108,95 @@ function Flaws({ flaws }: { flaws: FlawCheck[] }) {
   );
 }
 
-function BrokenLinks({ urls }: { urls: string[] }) {
-  const { data, error } = useSWR(
-    `/_redirects`,
-    async (url) => {
-      try {
-        const response = await fetch(url, {
-          method: "post",
-          body: JSON.stringify({ urls }),
-          headers: { "Content-Type": "application/json" },
-        });
-        if (!response.ok) {
-          throw new Error(`${response.status} on ${url}`);
-        }
-        return await response.json();
-      } catch (err) {
-        throw err;
-      }
-    },
-    {
-      // The chances of redirects to have changed since first load is
-      // just too small to justify using 'revalidateOnFocus'.
-      revalidateOnFocus: false,
+interface Link {
+  href: string;
+  line: number;
+  column: number;
+  suggestion: string | null;
+  nth: number;
+}
+
+function BrokenLinks({ doc, links }: { doc: Doc; links: Link[] }) {
+
+  // The `links` array will look something like this:
+  //  [
+  //    {href: 'foo', line: 12, ...},
+  //    {href: 'bar', line: 44, ...},
+  //    {href: 'foo', line: 53, ...},
+  //  ]
+  // So note that there are 2 'foo' in there. When we match this against
+  // the `document.querySelectorAll("a[href]")`, how are you supposed to
+  // which which of the 'foo' one you're referring to?
+  // This code, makes it so that each link in the array get's an 'nth'
+  // key too. It would, in our example, become:
+  //  [
+  //    {href: 'foo', nth: 0, line: 12, ...},
+  //    {href: 'bar', nth: 0, line: 44, ...},
+  //    {href: 'foo', nth: 1, line: 53, ...},
+  //  ]
+  // Now, when you've filtered the list of
+  // all `document.querySelectorAll("a[href]")` that matches 'foo'
+  // you know *which* one to pick.
+  const indexes = {};
+  links.forEach((link, i) => {
+    if (!(link.href in indexes)) {
+      indexes[link.href] = 0;
     }
-  );
+    link.nth = indexes[link.href];
+    indexes[link.href]++;
+  });
 
+  const [opening, setOpening] = React.useState<string | null>(null);
   useEffect(() => {
-    // 'data' will be 'undefined' until the server has had a chance to
-    // compute all the redirects. Don't bother until we have that data.
-    if (!data) {
-      return;
+    let unsetOpeningTimer: ReturnType<typeof setTimeout>;
+    if (opening) {
+      unsetOpeningTimer = setTimeout(() => {
+        setOpening(null);
+      }, 3000);
     }
+    return () => {
+      if (unsetOpeningTimer) {
+        clearTimeout(unsetOpeningTimer);
+      }
+    };
+  }, [opening]);
+
+  const filepath = doc.source.folder + "/index.html";
+
+  function openInEditor(key: string, line: number, column: number) {
```

Comment:
This function is very similar to the one for macros (in the same file) but not entirely the same. 
Do you think I should refactor it so that the function is generic and can work for both `Macros` and for `BrokenLinks`?